clients/include/utility.hpp

```diff
@@ -912,6 +912,122 @@ hipsparseStatus_t host_nnz(hipsparseDirection_t      dirA,
     return HIPSPARSE_STATUS_SUCCESS;
 }
 
+template <hipsparseDirection_t DIRA, typename T>
+hipsparseStatus_t host_dense2csx(int                  m,
+                                 int                  n,
+                                 hipsparseIndexBase_t base,
+                                 const T*             A,
+                                 int                  ld,
+                                 const int*           nnz_per_row_columns,
+                                 T*                   csx_val,
+                                 int*                 csx_row_col_ptr,
+                                 int*                 csx_col_row_ind)
+{
+    static constexpr T s_zero = {};
+    int                len    = (HIPSPARSE_DIRECTION_ROW == DIRA) ? m : n;
+    *csx_row_col_ptr          = base;
+    for(int i = 0; i < len; ++i)
+    {
+        csx_row_col_ptr[i + 1] = nnz_per_row_columns[i] + csx_row_col_ptr[i];
+    }
+
+    switch(DIRA)
+    {
+    case HIPSPARSE_DIRECTION_COLUMN:
+    {
+        for(int j = 0; j < n; ++j)
+        {
+            for(int i = 0; i < m; ++i)
+            {
+                if(A[j * ld + i] != s_zero)
+                {
+                    *csx_val++         = A[j * ld + i];
+                    *csx_col_row_ind++ = i + base;
+                }
+            }
+        }
+        return HIPSPARSE_STATUS_SUCCESS;
+    }
+
+    case HIPSPARSE_DIRECTION_ROW:
+    {
+        //
+        // Does not matter having an orthogonal traversal ... testing only.
+        // Otherwise, we would use csxRowPtrA to store the shifts.
+        // and once the job is done a simple memory move would reinitialize the csxRowPtrA to its initial state)
+        //
+        for(int i = 0; i < m; ++i)
+        {
+            for(int j = 0; j < n; ++j)
+            {
+                if(A[j * ld + i] != s_zero)
+                {
+                    *csx_val++         = A[j * ld + i];
+                    *csx_col_row_ind++ = j + base;
+                }
+            }
+        }
+        return HIPSPARSE_STATUS_SUCCESS;
+    }
+    }
+
+    return HIPSPARSE_STATUS_INVALID_VALUE;
```

Comment:
Looking at this though maybe it is unnecessary to return a status. I mean the only what you can get an invalid status is if you used a direction that was not row or column but those are the only options? I am ok with either having it return a status or void so its up to you and Nico

Reply:
let's bring this topic to the next meeting.